src/sync/mutex.rs

```diff
@@ -94,15 +94,79 @@ impl<T: ?Sized> Mutex<T> {
                 mutex: self,
             })),
             Err(TryLockError::WouldBlock) => panic!("mutex state out of sync"),
-        }
+        };
+
+        // We need to let other threads in here so they may fail a `try_lock`
```

Comment:
You could explain this a bit more, maybe even pointing to the `concurrent_lock_try_lock` test and saying that test doesn't fail the `try_lock`s if this context switch is not here.